src/consensus.rs

```diff
@@ -840,18 +925,98 @@ impl Consensus {
             // Broadcast the *Commit* Quorum Certificate to all validators
             self.broadcast_net_message(ConsensusNetMessage::Commit(
                 commit_quorum_certificate.clone(),
+                consensus_proposal_hash,
             ))?;
 
             // Process the same locally.
-            let consensus_proposal = std::mem::take(&mut self.bft_round_state.consensus_proposal);
-            self.commit_proposal(consensus_proposal, commit_quorum_certificate)?;
+            self.try_commit_current_proposal(commit_quorum_certificate)?;
         }
         // TODO(?): Update behaviour when having more ?
         Ok(())
     }
 
     /// Message received by follower.
-    fn on_commit(&mut self, commit_quorum_certificate: QuorumCertificate) -> Result<()> {
+    fn on_commit(
+        &mut self,
+        commit_quorum_certificate: QuorumCertificate,
+        proposal_hash_hint: ConsensusProposalHash,
+    ) -> Result<()> {
+        match self.bft_round_state.state_tag {
+            StateTag::Follower => {
+                return self.try_commit_current_proposal(commit_quorum_certificate)
+            }
+            StateTag::Joining => {}
+            _ => bail!("Commit message received while not follower"),
+        };
+
+        // We are joining consensus, try to sync our state.
+        // First find the prepare message to this commit.
+        let Some(proposal_index) = self
+            .bft_round_state
+            .joining
+            .buffered_prepares
+            .iter()
+            .position(|p| p.hash() == proposal_hash_hint)
+        else {
+            // Maybe we just missed it, carry on.
+            return Ok(());
+        };
+        // Use that as our proposal.
+        self.bft_round_state.consensus_proposal = self
+            .bft_round_state
+            .joining
+            .buffered_prepares
+            .swap_remove(proposal_index);
+        self.bft_round_state.joining.buffered_prepares.clear();
+
+        // At this point check that we're caught up enough that it's realistic to verify the QC.
+        if self.bft_round_state.joining.staking_updated_to + 1
+            < self.bft_round_state.consensus_proposal.slot
+        {
```

Comment:
```suggestion
        if self.bft_round_state.joining.staking_updated_to + 1
            <= self.bft_round_state.consensus_proposal.slot
        {
```
Si on a rettrapé jusqu'à `n` et que le slot est `n+1` on ne va pas prendre en compte le commit alors qu'on pourra le faire. non ?

Reply:
non justement mon check est bon pour ça -> On a rattrapé jusque n, donc `n + 1 < n + 1` est faux donc on ignore pas